### ISSUE TITLE
Attach multiple decorators

### ISSUE THREAD

**Issue by browniefed** [ractivejs/ractive#399]
I was under the assumption that you could attach more than one decorator.
We are attempting to have 2 decorators. One for editing, and one for validating.

Although it appears that you can only have one decorator per element.
So I guess this is a feature request to have more than one decorator per element.

http://jsfiddle.net/M3MkH/

**Comment by briancray** [comment]
Just searched for this after I ran into the same issue. Perhaps a format of `decorator-name="options"` and if no options are needed, a simple boolean attribute like, `<input decorator-autofill>` where name is the exported name of the decorator.

**Comment by akhoury** [comment]
I second @briancray 's suggestion - the `decorator-{decoratorName}` format kind of matches the `on-{eventName}` convention.

**Comment by LeJared** [comment]
+1 for ```decorator-name="options"```

**Comment by Madgvox** [comment]
Although this is a relatively old topic, I have to agree that this proposed change is at least worth discussing. Perhaps this can be readdressed after the 0.8 release.

**Comment by LeJared** [comment]
Until the use of multiple decorators is supported by Ractive natively, I've create a "Meta"-Decorator, for combining multiple decorators on one element.

In contrast to [Ractive-decorator-helpers](http://martypdx.github.io/Ractive-decorator-helpers/)  you  can use my solution out of the box with your existing decorators, because you don't have to change the way to write your decorators (which I don't like about *Ractive-decorator-helpers*).

Usage
```html
<div decorator="combine:{decorator1:[param1, param2], decorator2:param, decorator3:{{myValue}}}"> Text </div>
```
Nothing els to do!

Here it is:
```javascript
/**
 * Meta-Decorator to use multiple decorators on a single HTML-element
 * usage: decorator="combine:{decorator1:[param1, param2], decorator2:param}"
 */
Ractive.decorators.combine = function (node, data) {
	var decoratorName, decorator, parameters;
	var instance = this;
	var decorators = data || {};

	for (decoratorName in decorators) {
		if ((decorator = instance.decorators[decoratorName])) {
			parameters = makeArray(decorators[decoratorName]);
			parameters.unshift(node);
			decorators[decoratorName] = decorator.apply(instance, parameters);
		} else {
			delete decorators[decoratorName];
		}
	}

	return {
		teardown: function () {
			var decoratorName, decorator;

			for (decoratorName in decorators) {
				decorator = instance.decorators[decoratorName];
				decorator.teardown && decorator.teardown();
			}
		},
		update: function (data) {
			var decoratorName, decorator, parameters;

			for (decoratorName in decorators) {
				decorator = decorators[decoratorName];
				parameters = makeArray(data[decoratorName]);
				decorator.update && decorator.update.apply(instance, parameters);
			}
		}
	};
};
```

**Comment by heavyk** [comment]
@LeJared  brilliant!!

**Comment by akhoury** [comment]
@LeJared how is that different from the `multi` decorator mentioned a few comments above ? By @martypdx ?

**Comment by LeJared** [comment]
To use the *multi* decorate from above or *Ractive-decorator-helpers*  you have to change the way you create and assign your decorators to your Ractive instance. I don't like that, because you can't use exiting decorators out of the box, but have to create special combined versions of your decorators, to use them togethe. This also means, You have to create as create a new *"combined decorator"* for *every* combination separately!

Using my solution, you only have to change the way you assign your decorators in then templates and you can use any combinations of any decorators on the fly by just adding them in the template.

Benefit: once, Ractive supports multiple decorators natively, you only have to change your templates (you'll have to do this anyway!) but you don't need to change your decorators them selfs.

**BTW:** I've fixed a bug in the code above.

**Comment by akhoury** [comment]
@LeJared 
> you have to change the way you create and assign your decorators to your Ractive instance

I guess the initial idea of @martypdx that was true, but see my comment above, which basically was a slight improvement 
https://github.com/ractivejs/ractive/issues/399#issuecomment-64007822

that's how you use it in the template - all other decorators remain the same, and you don't have to create a combination before hand.

```
<div 
    id="myelement"
    decorator="multi:{ 
        localdecorator: [ {{ dynamicArg1 }}, arg2, arg3 ], 
        globaldecorator: {{singleArg}}, 
        anotherdecorator: true 
    }"
></div>
```

**Comment by martypdx** [comment]
Yeah, after I wrote that I realized that you could just look them up from global registry. 

What I find interesting is that I have yet to need to double up decorators on an element.

**Comment by Madgvox** [comment]
> What I find interesting is that I have yet to need to double up decorators on an element.

The only situation where I've needed to double up a decorator is for a 'truncate' decorator, where I also want to have a 'tooltip' decorator on the element. I didn't need to have a multi-decorator plugin, though. I just instantiated the tooltip decorator from the truncate decorator manually. I'm not sure if that strategy is generally applicable however.

**Comment by LeJared** [comment]
@martypdx & @akhoury: Sorry, I missed that. So then both solutions are doing essentially the same.

@martypdx: What do you mean by *have yet to need to double up decorators on an element*? (Sorry, I'm not native english speaker).

**Comment by martypdx** [comment]
> What do you mean by have yet to need to double up decorators on an element? 

@LeJared:  I haven't needed a multi decorator.

**Comment by JonDum** [comment]
I slapped up @akhoury's solution as a repo and put it up on npm as `ractive-multi-decorator`

**Comment by martypdx** [comment]
@JonDum :+1: 

Maybe it should be `ractive-each-decorator`:

```html
<div 
    id="myelement"
    decorator="each:{ 
        localdecorator: [ {{ dynamicArg1 }}, arg2, arg3 ], 
        globaldecorator: {{singleArg}}, 
        anotherdecorator: true 
    }"
></div>
```

**Comment by zaus** [comment]
Use case for multiple decorators:  I have a list of items that I want to be able to both drag-drop AND click a button to add or remove items from the list.  Would use:
* [Sortable](http://ractivejs.github.io/Ractive-decorators-sortable/)
* [Addable](https://github.com/zaus/Ractive-decorators-addable) _(in progress)_

**Comment by martypdx** [comment]
@zaus I'm using @JonDum's repo too and works great.

**Comment by akhoury** [comment]
updated @JonDum's [repo](https://github.com/JonDum/ractive-multi-decorator/commit/203bd2c18fd8535f527aecf1a86525082e6844b9) to avoid unnecessary updates on decorators which arguments haven't changed.

**Comment by akhoury** [comment]
hey @browniefed, may i ask why is this being closed? I thought it was labeled as a future enhancement.
Has it been decided that this feature won't be supported in core?

**Comment by browniefed** [comment]
@akhoury ah sorry I assumed it was already implemented for some reason haha, was cleaning up all my old open issues. Will keep thsi open

**Comment by evs-chris** [comment]
This now possible in edge with multiple `as-${decorator}` directives 😄

**Comment by TehShrike** [comment]
BOOYAH.  What version will it be released in?  Is Edge 0.8, 0.9, or even farther out?

**Comment by evs-chris** [comment]
Edge is 0.8. I'm planning on cutting an alpha week after next.
